### PR TITLE
Merge uOttawa parser

### PR DESCRIPTION
**Pull request by sarmadhashmi**
Merging new uOttawa parser #7

### CODE REVIEW

File: parsers/uottawa-parser-2.js [hunk 1]
```diff
@@ -25,14 +24,20 @@ function parse(text) {
...
-      obj['semester_end'] = Date.parse(semester[2] + ' ' + semester[3]);
+      obj['course_code_faculty'] = course[1];
+      obj['course_code_number'] = course[2];
+      obj['section'] = course[3];
+      obj['course_name'] = course[4];
+      obj['professor'] = professor[0];
+      var sem_start = new Date(semester[1] + ' ' + semester[3]); 
+      var sem_end = new Date(semester[2] + ' ' + semester[3]); 
+      obj['semester_start'] = {'year': sem_start.getFullYear(), 'month': sem_start.getMonth(), 'day': sem_start.getDate()};
+      obj['semester_end'] = {'year': sem_end.getFullYear(), 'month': sem_end.getMonth(), 'day': sem_end.getDate()};
```
Comment (AbeOmor) [inline parsers/uottawa-parser-2.js]: Are you going to account for date format like I did?

File: parsers/uottawa-parser-2.js [hunk 2]
```diff
@@ -25,14 +24,20 @@ function parse(text) {
...
       obj['where'] = where[0];
       obj['day'] = time[1]; 
-      obj['start_time'] = time[2];
-      obj['end_time'] = time[3];
-      obj['class_type'] = time[4];
+      obj['start_time'] = {'hour': parseInt(time[2], 10), 'minute': parseInt(time[3], 10)};
+      obj['end_time'] = {'hour': parseInt(time[4], 10), 'minute': parseInt(time[5], 10)};
+      obj['class_type'] = time[6];
```
Comment (AbeOmor) [inline parsers/uottawa-parser-2.js]: Are you going to account for the 24HR and 12HR clock like I did?

File: parsers/uottawa-parser-2.js [hunk 3]
```diff
@@ -1,11 +1,10 @@
...
-var _courseRegex = /\b[A-Z]{3}[1-9]{1}[0-9]{3}[A-Z]?\b/;
+var _courseRegex = /\b([A-Z]{3})([1-9]{1}[0-9]{3})([A-Z])?\s+(\b([A-Z.]+\s([A-Z.&]+\s?)+)\b)/;
 var _courseRegexLookAhead = /(?=\b[A-Z]{3}[1-9]{1}[0-9]{3}[A-Z]?\b)/g;
 var _timeRegexLookAhead = /(?=\b(Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday)\s+[0-9]{1,2}:[0-9]{2}\s+to\s+[0-9]{2}:[0-9]{2}\s+[A-Za-z]*\b)/g;
-var _timeRegex = /\b(Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday)\s+([0-9]{1,2}:[0-9]{2})\s+to\s+([0-9]{2}:[0-9]{2})\s+([A-Za-z]*)\b/;
+var _timeRegex = /\b(Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday)\s+([0-9]{1,2}):([0-9]{2})\s+to\s+([0-9]{2}):([0-9]{2})\s+([A-Za-z]*)\b/;
 var _professorRegex = /\b[A-Za-z\-.]+\s(?:[A-Za-z\-.]+\s?)+\b/;
 var _semesterRegex = /([0-9]{2}\s[A-Z]{1}[a-z]{2,8})\sto\s([0-9]{2}\s[A-Z]{1}[a-z]{2,8})\s([0-9]{4})/; 
-
 function parse(text) {
```
Comment (AbeOmor) [inline parsers/uottawa-parser-2.js]: Comment the function like:
```
/**
 * Description..
 * @param 
 * @return 
 */
```

File: parsers/uottawa-parser-2.js [hunk 4]
```diff
@@ -25,14 +24,20 @@ function parse(text) {
...
       obj['day'] = time[1]; 
-      obj['start_time'] = time[2];
-      obj['end_time'] = time[3];
-      obj['class_type'] = time[4];
+      obj['start_time'] = {'hour': parseInt(time[2], 10), 'minute': parseInt(time[3], 10)};
+      obj['end_time'] = {'hour': parseInt(time[4], 10), 'minute': parseInt(time[5], 10)};
+      obj['class_type'] = time[6];
       timetable.push(obj);
     }
   }
```
Comment (AbeOmor) [inline parsers/uottawa-parser-2.js]: whitespace after this closed bracket

File: parsers/uottawa.js [hunk 5]
```diff
@@ -1,183 +1,77 @@
...
+          'day': sem_start.getDate()
+        },
+        'semester_end': {
+          'year': sem_end.getFullYear(),
+          'month': sem_end.getMonth(),
+          'day': sem_end.getDate()
+        },
+        'where': where[0],
+        'day': time[1],
+        'start_time': {
```
Comment (AbeOmor) [inline parsers/uottawa.js]: Are you going to account for date format or time format 12HR and 24HR like I did?

Comment (sarmadhashmi) [inline parsers/uottawa.js]: As of right now, I don't see anyway to change the time in our course schedules from 24 HR to 12 HR. If that happens, I'll update it but for now it's not worth the effort.

File: parsers/uottawa.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ var fs = require('fs')
   });
 
 // Relevant regular expressions
-var _locationRegex = /\b[A-Z]{3,5}\s[A-Z0-9]{2,4}\b/;
+var _locationRegex = /\b[A-Z]{3,5}\s[A-Z0-9]{2,6}\b/;
```
Comment (AbeOmor) [inline parsers/uottawa.js]: What if my class name is SEE CLASS NOTES

Comment (sarmadhashmi) [inline parsers/uottawa.js]: `var _courseRegex = /\b([A-Z]{3})([1-9]{1}[0-9]{3})([A-Z])?\s+(\b([A-Z.]+\s([A-Z.&]+\s?)+)\b)/;`

first set of () is for the course code, second is for the course name, which is an arbitrary number of capital letters followed by a space and more capital letters, so on and so forth. "SEE CLASS NOTES" would be supported.

File: parsers/uottawa.js [hunk 7]
```diff
@@ -37,7 +38,7 @@ function parse(text) {
       var professor = _professorRegex.exec(classes[j]);
       var time = _timeRegex.exec(classes[j]);
       var where = _locationRegex.exec(classes[j]);
-      if (!professor || !time || !where) continue;
```
Comment (AbeOmor) [inline parsers/uottawa.js]: Some this means there are some calendar with no input of prof or location? Isn't there a default name

Comment (sarmadhashmi) [inline parsers/uottawa.js]: Nope, it's left empty because they haven't been assigned yet. Usually happens 2-3 weeks before semester start.

**Pull request merged**